dclock.man
.TH DCLOCK 1 "1 March 1988" "X Version 11"
.SH NAME
dclock - digital clock for X
.SH SYNOPSIS
.B dclock
[-\fItoolkitoption\fP ...] [-option ...]
.SH DESCRIPTION
The
.I dclock 
program displays the time in digital format only.  The time is
updated on a per second basis or on a per minute basis.  This program is
nothing more than a wrapper around the dclock widget not associated with
any particular widget set.
.sp
When the clock is running, the user may change attributes by typing:
.in +2
.ta 2i
.nf
\fBr\fP	Toggles \fBReverse Video\fP.
\fBs\fP	Toggles the \fBseconds\fP display.
\fBb\fP	Toggles the \fBbell\fP attribute.
\fBj\fP	Toggles the \fBjump/scroll\fP attribute.
\fBf\fP	Toggles the \fBfade\fP attribute.
\fBd\fP	Toggles the \fBdate\fP format.
\fBm\fP	Toggles the \fBmilitary time\fP format.
\fBa\fP	Toggles the \fBalarm clock\fP.
\fBt\fP	Toggles the \fBtails\fP attribute.
\fBq\fP	quit the program.
.sp
.fi
.in -2
The third mouse button puts the clock in the mode where the alarm clock can
be set.  The alarm must be set in 24-hour format to distringuish between
am and pm time.  Digits can be changed by selecting the digit with the
left or middle mouse button.  The Left button advances the time while the
Middle button moves the time backwards.  Using the left or middle button
over the text at the bottom of the clock toggles the alarm from actually
being set (or, use the 'a' key).
.SH OPTIONS
.I Dclock
accepts all of the standard X Toolkit command line options along with the 
additional options listed below:
.TP 8
.B \-help
This option indicates that a brief summary of the allowed options should be
printed on the standard error.
.TP 8
.B \-bell
This option indicates that the bell will beep
once on the half hour and twice on the hour.
.TP 8
.B \-miltime
This option causes the clock to display the time in 24 hour (military)
time format.
.TP 8
.B \-tails
This option puts "tails" on the digits 6 and 9.
.TP 8
.B \-scroll
.TP 8
.B \-noscroll
.br
When the time changes, the digits scroll from the previous digit to the
next digit.  Since this is on by default, the -noscroll option can turn
it off.
.TP 8
.B \-fade
This option causes the display to fade from one digit to the next instead
of jumping or scrolling.  If fade is specified, the \fBscroll\fP attribute
will be forced to off.
.TP 8
.B \-fadeRate \fIn\fP
The fadeRate attribute takes an integer value which specifies the
number of milliseconds to wait between iterations when fading from
one digit to the next.  A larger value will make the fade slower and
more pronounced.  The default is 50.
.TP 8
.B \-date \fI"format"\fP
The date is printed under the time in the specified font.  The string
displayed is in the "format" argument.  If the string contains a formatting
character (%), then the next character is examined and a value is inserted
into the string.  Example:
.br
.sp
.ti +2
dclock -date "Today is %W"
.sp
The date string will print "Today is Friday" if the weekday name happens
to be friday.  The formatting characters that are understood are:
.in +2
.nf
%W	Full weekday name
%w	Three-char weekday name (Sun, Mon, Tue...)
%M	Full month name
%m	Three-char month name (Jan, Feb, Mar...)
%d	The date (numerical day number of the month)
%Y	Full year (4 digits)
%y	2-digit year number
.fi
.in -2
.TP 8
.B \-alarm
.TP 8
.B \-noalarm
.br
The alarm is turned on or off.  Alarm rings bell and toggles the
reverse video five times.
.br
.TP 8
.B \-alarmTime \fIHH:MM\fP
If alarm is set, it goes off at time specified in 24-hour format.
.TP 8
.B \-seconds
This option will update the clock every second and display the time
including the seconds.
.TP 8
.B \-bg \fIcolor\fP
This option specifies the color to use for the background of the window.  
The default is ``white.''
.TP 8
.B \-bd \fIcolor\fP
This option specifies the color to use for the border of the window.
The default is ``black.''
.TP 8
.B \-bw \fInumber\fP
This option specifies the width in pixels of the border surrounding the window.
.TP 8
.B \-fg \fIcolor\fP
This option specifies the color to use for displaying text.  The default is 
``black''.
.TP 8
.B \-fn \fIfont\fP
This option specifies the font to be used for displaying normal text.  The
default is ``Fixed.''
.TP 8
.B \-rv
This option indicates that reverse video should be simulated by swapping
the foreground and background colors.
.TP 8
.B \-geometry \fIgeometry\fP
This option specifies the prefered size and position of the clock window.
.TP 8
.B \-display \fIhost\fP:\fIdisplay\fP
This option specifies the X server to contact.
.TP 8
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.  This is especially
useful for setting resources that do not have separate command line options.
.SH X DEFAULTS
It understands all of the core resource names and
classes as well as:
.PP
.TP 8
.B width (\fPclass\fB Width)
Specifies the width of the clock.
.TP 8
.B height (\fPclass\fB Height)
Specifies the height of the clock.
.TP 8
.B foreground (\fPclass\fB Foreground)
Specifies the color for the tic marks.  Using the class specifies the
color for all things that normally would appear in the foreground color.
The default is ``black'' since the core default for background is ``white.''
.TP 8
.B bell (\fPclass\fB Boolean)
Specifies whether or not a bell should be rung on the hour and half hour.
.TP 8
.B font (\fPclass\fB Font)
Specifies the font to be used for the date.
.TP 8
.B miltime (\fPclass\fB Boolean)
Specifies whether the military (24hr) time format should be used.
.TP 8
.B reverseVideo (\fPclass\fB ReverseVideo)
Specifies that the foreground and background colors should be reversed.
.TP 8
.B scroll (\fPclass\fB Boolean)
Specifies whether the digits should scroll or not.
.TP 8
.B fade (\fPclass\fB Boolean)
Specifies whether the digits should fade or not.
.TP 8
.B fadeRate (\fPclass\fB Time)
The number of milliseconds of delay between iterations of fading.
.TP 8
.B seconds (\fPclass\fB Boolean)
Specifies whether the seconds should be displayed or not.
.TP 8
.B bell (\fPclass\fB Boolean)
Specifies whether the bell should sound on the half hour and on the hour.
.TP 8
.B tails (\fPclass\fB Boolean)
Specifies whether tails should be put on the digits 6 and 9.
.TP 8
.B alarm (\fPclass\fB Boolean)
Specifies whether the alarm should go off at the specified time (alarmTime).
.TP 8
.B alarmTime (\fPclass\fB String)
Specifies the time alarm goes off if alarm is set.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH "SEE ALSO"
X(1), xrdb(1), time(3C).
.SH BUGS
.I Dclock
believes the system clock.
.PP
Scrolling from 9 to 10 O'Clock seems weird, but chances are you won't
notice it.
.PP
If the window is too small, the seconds are tough to read.  This should
be recognized by the program and should display seconds using a font
rather than the bitmaps used by the clock itself.
.SH COPYRIGHT
Copyright (c) 1988, Dan Heller.
.SH AUTHOR
.nf
Dan Heller -- <argv@sun.com> or <dheller@cory.berkeley.edu>
Fade & tails by Steve Reinhardt, <steven@fts.webo.dg.com>
.fi


dclock.c
/*
 * Copyright (c) 1988 Dan Heller <argv@sun.com>
 * dclock -- program to demonstrate how to use the digital-clock widget.
 * To specify a date, the date format is a string of characters.  If a
 * character in the string is a %-sign, the next character is examined
 * and a value is inserted into the string.  Example:
 *    dclock -date "Today is %W"
 * The date string will print "Today is" and then the %W will be replaced
 * by the current weekday name.  The parameters are:
 *    %W	full weekday name
 *    %w	three-char weekday name (sun, mon, tue, wed...)
 *    %M	full month name
 *    %m        three-char abbreviation for that month (jan, feb, mar...)
 *    %d	The date (numerical number of the month)
 *    %Y	full year (4 digits)
 *    %y	2-digit year number
 *
 * To specify seconds to be displayed, use "-seconds" or use the resource
 * manager: *Dclock.seconds: on
 */
#include <stdio.h>
#include <X11/Intrinsic.h>
#include "Dclock.h"

static XrmOptionDescRec options[] = {
    {"-date",	 "*Dclock.date",	XrmoptionSepArg, NULL },
    {"-seconds", "*Dclock.seconds",	XrmoptionNoArg, "TRUE" },
    {"-miltime", "*Dclock.miltime",	XrmoptionNoArg, "TRUE" },
    {"-bell",	 "*Dclock.bell",	XrmoptionNoArg, "TRUE" },
    {"-scroll",  "*Dclock.scroll",	XrmoptionNoArg, "TRUE" },
    {"-noscroll","*Dclock.scroll",	XrmoptionNoArg, "FALSE" },
    {"-alarm",   "*Dclock.alarm",	XrmoptionNoArg, "TRUE" },
    {"-noalarm",   "*Dclock.alarm",	XrmoptionNoArg, "FALSE" },
    {"-alarmTime","*Dclock.alarmTime",	XrmoptionSepArg, NULL },
    {"-tails",	 "*Dclock.tails",	XrmoptionNoArg, "TRUE" },
    {"-fade",	 "*Dclock.fade",	XrmoptionNoArg, "TRUE" },
    {"-fadeRate", "*Dclock.fadeRate",	XrmoptionSepArg, NULL },
};

static void
Usage(name, args)
String name, *args;
{
    static char *help_message[] = {
	"where options include:",
	"    -bg color			background color",
	"    -fg color			foreground color",
	"    -fn font			font name",
	"    -geometry geom		size of mailbox",
	"    -display host:dpy		X server to contact",
	"    -seconds [on/off]		display seconds",
	"    -miltime [on/off]		display time in 24 hour format",
	"    -bell [on/off]		ring bell each half hour",
	"    -scroll [on/off]		turn on/off scrolling of numbers",
	"    -date \"date format\"	show the date in specified format",
	"    -alarm [on/off]		turn on/off alarm",
	"    -alarmTime hh:mm		Time alarm goes off",
	"    -tails			put tails on 6 and 9",
	"    -fade			fade numbers",
	"    -fadeRate			wait between fade steps (in msec)",
	NULL
    };
    char **cpp;

    fprintf(stderr, "Invalid Args:");
    while (*args)
	fprintf(stderr, " \"%s\"", *args++);
    fputc('\n', stderr);
    fprintf(stderr, "usage: %s [-options ...]\n", name);
    for (cpp = help_message; *cpp; cpp++)
	fprintf(stderr, "%s\n", *cpp);
    exit(1);
}

static void
quit()
{
    exit(0);
}

static XtActionsRec actionsList[] = {
    { "quit",	quit },
};

main(argc, argv)
char *argv[];
{
    Widget toplevel, clock_w;
    char *name, *rindex();

    if (name = rindex(argv[0], '/'))
	name++;
    else
	name = argv[0];

    toplevel = XtInitialize(name, "DClock", options, XtNumber(options),
			    &argc, argv);
    XtAddActions(actionsList, 1);

    if (argc != 1)
	Usage(name, argv+1);

    clock_w = XtCreateManagedWidget(name, dclockWidgetClass, toplevel, NULL, 0);
    XtOverrideTranslations(clock_w, XtParseTranslationTable("<Key>q: quit()"));

    XtRealizeWidget(toplevel);
    XtMainLoop();
}
